src/histogram.rs

```diff
@@ -0,0 +1,332 @@
+    // e.g. if the revision of the running binary is put into a
+    // label. Another, more advanced purpose is if more than one Collector
+    // needs to collect Metrics with the same fully-qualified name. In that
+    // case, those Metrics must differ in the values of their
+    // ConstLabels. See the Collector examples.
+    //
+    // If the value of a label never changes (not even between binaries),
+    // that label most likely should not be a label at all (but part of the
+    // metric name).
+    pub const_labels: HashMap<String, String>,
+
+    // buckets defines the buckets into which observations are counted. Each
+    // element in the slice is the upper inclusive bound of a bucket. The
+    // values must be sorted in strictly increasing order. There is no need
+    // to add a highest bucket with +Inf bound, it will be added
+    // implicitly. The default value is DefBuckets.
+    pub buckets: Vec<f64>,
+}
+
+impl HistogramOpts {
+    pub fn new<S: Into<String>>(name: S, help: S) -> HistogramOpts {
+        HistogramOpts {
+            namespace: "".to_owned(),
+            sub_system: "".to_owned(),
+            name: name.into(),
+            help: help.into(),
+            const_labels: HashMap::new(),
+            buckets: Vec::from(DEFAULT_BUCKETS as &'static [f64]),
+        }
+    }
+
+    pub fn namespace<S: Into<String>>(mut self, namesapce: S) -> Self {
+        self.namespace = namesapce.into();
+        self
+    }
+
+    pub fn sub_system<S: Into<String>>(mut self, sub_system: S) -> Self {
+        self.sub_system = sub_system.into();
+        self
+    }
+
+    pub fn const_labels(mut self, labels: HashMap<String, String>) -> Result<Self> {
+        for k in labels.keys() {
+            if k == BUCKET_LABEL {
+                return Err(Error::Msg("`le` is not allowed as label name in histograms"
+                    .to_owned()));
+            }
+        }
+
+        self.const_labels = labels;
+        Ok(self)
+    }
+
+    pub fn const_label<S: Into<String>>(mut self, name: S, value: S) -> Result<Self> {
+        let name = name.into();
+        if name == BUCKET_LABEL {
+            return Err(Error::Msg("`le` is not allowed as label name in histograms".to_owned()));
+        }
+
+        self.const_labels.insert(name, value.into());
+        Ok(self)
+    }
+
+    pub fn buckets(mut self, buckets: Vec<f64>) -> Result<Self> {
+        // TODO: check buckets order.
+        self.buckets = buckets;
+        Ok(self)
+    }
+
+    pub fn fq_name(&self) -> String {
+        build_fq_name(&self.namespace, &self.sub_system, &self.name)
+    }
+}
+
+#[derive(Debug)]
+struct HistogramCore {
+    sum: f64,
+    count: u64,
+
+    upper_bounds: Vec<f64>,
+    counts: Vec<u64>,
+}
+
+impl HistogramCore {
+    fn new() -> HistogramCore {
+        HistogramCore::with_buckets(vec![]).unwrap()
+    }
+
+    fn with_buckets(mut buckets: Vec<f64>) -> Result<HistogramCore> {
```

Comment:
check buckets invalidation
1. must be in increasing order. 
2. check +Inf in the end